arrows/vxl/hashed_image_classifier.cxx

```diff
@@ -0,0 +1,518 @@
+
+  for( unsigned j = 0; j < output_image.nj(); ++j )
+  {
+    weight_t* dpixel = &output_image( 0, j );
+
+    for( unsigned f = 0; f < features; ++f )
+    {
+      spixels[ f ] = &input_features[ f ]( 0, j );
+    }
+
+    for( unsigned i = 0; i < output_image.ni(); ++i, dpixel += distep )
+    {
+      for( unsigned f = 0; f < features; ++f )
+      {
+        *dpixel += feature_weights[ f ][ *spixels[ f ] ];
+        spixels[ f ] += sisteps[ f ];
+      }
+    }
+  }
+}
+
+// ----------------------------------------------------------------------------
+// Classify some chain of hashed input images, but only on specific pixels
+// as decided by the given mask
+template < typename FeatureType, typename OutputType >
+void
+hashed_image_classifier< FeatureType, OutputType >
+::classify_images( feature_vector_t const& input_features,
+                   mask_image_t const& mask,
+                   weight_image_t& output_image,
+                   weight_t const offset ) const
+{
+  classify_images( &input_features[ 0 ],
+                   input_features.size(),
+                   mask,
+                   output_image,
+                   offset );
+}
+
+// ----------------------------------------------------------------------------
+// Classify some chain of hashed input images, but only on specific pixels
+// as decided by the given mask
+template < typename FeatureType, typename OutputType >
+void
+hashed_image_classifier< FeatureType, OutputType >
+::classify_images( input_image_t const* input_features,
+                   unsigned const features,
+                   mask_image_t const& mask,
+                   weight_image_t& output_image,
+                   weight_t const offset ) const
+{
+  if( !model_->is_valid() )
+  {
+    std::cerr << "Internal classifier is invalid" << std::endl;
+  }
+  if( features != model_->num_features )
+  {
+    std::cerr << "Feature counts don't match" << std::endl;
+  }
+
+  output_image.set_size( input_features[ 0 ].ni(), input_features[ 0 ].nj() );
+
+  weight_t** const feature_weights = &model_->feature_weights[ 0 ];
+
+  for( unsigned j = 0; j < output_image.nj(); ++j )
+  {
+    for( unsigned i = 0; i < output_image.ni(); ++i )
+
+    {
+      if( mask( i, j ) )
+      {
+        weight_t& output = output_image( i, j );
+
+        output = offset;
+
+        for( unsigned f = 0; f < features; ++f )
+        {
+          output += feature_weights[ f ][ input_features[ f ]( i, j ) ];
+        }
+      }
+    }
+  }
+}
+
+// ----------------------------------------------------------------------------
+// Load a model from file
+template < typename FeatureType, typename OutputType >
+bool
+hashed_image_classifier< FeatureType, OutputType >
+::load_from_file( std::string const& file )
+{
+  model_.reset( new model_t() );
+
+  // Open model file for reading
+  std::ifstream input( file.c_str() );
+
+  if( !input.is_open() )
+  {
+    std::cout << "Unable to open input file: " << file << std::endl;
```

Comment:
Thinking these logging messages should be hooked into the kwiver logging rather than std.

Reply:
I'm guessing I'm missing something obvious but I've been struggling for a while to find an example of kwiver logging outside of an arrow implementation. Is there a preferred method for instantiating a logger in non-arrow helper functions/classes?

Reply:
Sorry, I didn't realize this was not a part of the arrow. Disregard for now.

Reply:
No, we should address this.  Whether or not its part of an arrow class we should not use `cout` and use vital logger instead.  You don't need to have a log handle that is stored as part of an arrow class.  That is only for convenience and efficiency.  You can make a log handle on the spot right before you use it.  I'll see if I can find an example to point you at.  I don't recall the syntax.

Reply:
Here is an example:
https://github.com/Kitware/kwiver/blob/bc695ae841955884407d6320124067a72e3a8c3e/arrows/vxl/estimate_similarity_transform.cxx#L85-L86